sherlock.h

```diff
@@ -235,11 +241,31 @@ class StreamInstanceImpl {
   void operator=(StreamInstanceImpl&&) = delete;
 };
 
+// TODO(dkorolev): Move into Bricks/util/ ?
+template <typename B, typename E>
+struct can_be_stored_in_unique_ptr {
+  static constexpr bool value = false;
+};
+
+template <typename B, typename E>
+struct can_be_stored_in_unique_ptr<std::unique_ptr<B>, E> {
+  static constexpr bool value = std::is_same<B, E>::value || std::is_base_of<B, E>::value;
+};
+
 template <typename T>
 struct StreamInstance {
   StreamInstanceImpl<T>* impl_;
   inline explicit StreamInstance(StreamInstanceImpl<T>* impl) : impl_(impl) {}
+
   inline void Publish(const T& entry) { impl_->Publish(entry); }
+
+  // TODO(dkorolev): Add a test for this code.
+  // TODO(dkorolev): Perhaps eliminate the copy.
+  template <typename E>
+  typename std::enable_if<can_be_stored_in_unique_ptr<T, E>::value>::type Publish(const E& polymorphic_entry) {
+    impl_->PublishPolymorphic(new E(polymorphic_entry));
```

Comment:
I'm not so sure, to be honest: One would still be necessary, unless I configure the stream itself to not own its objects.

Certainly not looking into it this coming weel. Demo + syntax nits + FNCAS release + Sherlock persistence come before copy optimization.